apps/remote_control/lib/lexical/remote_control/build/state.ex

```diff
@@ -14,33 +14,54 @@ defmodule Lexical.RemoteControl.Build.State do
 
   use RemoteControl.Progress
 
-  defstruct project: nil, build_number: 0, uri_to_source_and_edit_time: %{}
+  defstruct project: nil,
+            build_number: 0,
+            uri_to_document: %{},
+            project_compile: :none
 
   def new(%Project{} = project) do
     %__MODULE__{project: project}
   end
 
-  def on_tick(%__MODULE__{} = state) do
-    {new_state, compiled_uris} =
-      Enum.reduce(state.uri_to_source_and_edit_time, {state, []}, fn
-        {uri, {document, edit_time}}, {state, compiled_uris} ->
-          if should_compile?(edit_time) do
-            new_state = increment_build_number(state)
-            compile_file(new_state, document)
-            {new_state, [uri | compiled_uris]}
-          else
-            {state, compiled_uris}
-          end
-      end)
+  def on_timeout(%__MODULE__{} = state) do
+    case state.project_compile do
+      :none -> :ok
+      :force -> compile_project(state, true)
+      :normal -> compile_project(state, false)
+    end
+
+    new_state =
+      state.uri_to_document
+      |> Map.values()
```

Comment:
Optional: can get rid of this and just do `Enum.reduce(state.uri_to_document, fn {_uri, document}, state -> ... end)`